src/server_manager/web_app/ui_components/outline-per-key-data-limit-dialog.ts

```diff
@@ -0,0 +1,380 @@
+          </paper-listbox>
+        </paper-dropdown-menu>
+      </div>
+    `;
+  }
+
+  private _queryAs<T extends HTMLElement>(selector: string): T {
+    return this.shadowRoot.querySelector(selector) as T;
+  }
+
+  private get _input(): PaperInputElement {
+    return this._queryAs<PaperInputElement>('#dataLimitInput');
+  }
+
+  private _dataLimitValue() {
+    return Number(this._input?.value) ?? 0;
+  }
+
+  private _dataLimitType() {
+    return this._queryAs<PaperListboxElement>('#unitsListbox').selected as 'GB' | 'MB';
+  }
+
+  private _getInternationalizedUnit(bytes: number) {
+    return formatBytesParts(bytes, this._language).unit;
+  }
+
+  private _initialUnit(): 'GB'|'MB' {
+    return bytesToDisplayDataAmount(this._activeDataLimit())?.unit || 'GB';
+  }
+
+  private _initialValue() {
+    return bytesToDisplayDataAmount(this._activeDataLimit())?.value.toString() || '';
+  }
+
+  private _activeDataLimit(): number|undefined {
+    // Returns the limit which currently is enforced on this key, or undefined if there is none.
+    return this._keyDataLimitBytes ?? this._serverDefaultLimitBytes;
+  }
+
+  private async _setCustomLimitTapped() {
+    this._showMenu = !this._showMenu;
+    if (this._showMenu) {
+      await this.updateComplete;
+      this._input?.focus();
+    }
+  }
+
+  private _setSaveButtonDisabledState() {
+    this._enableSave = !this._input?.invalid ?? false;
+  }
+
+  private _sendSaveEvent() {
+    const change = this._dataLimitChange();
+    const eventName = `${change === Change.REMOVED ? 'Remove' : 'Save'}PerKeyDataLimitRequested`;
+
+    this.dispatchEvent(new CustomEvent(eventName, {
+      // The target of this event is appRoot, not the dialog.  We send the full dialog element so
+      // that the app can control the dialog accordingly to user input.
+      detail: {dialog: this},
+      // Required for the event to bubble past a shadow DOM boundary
+      bubbles: true,
+      composed: true,
+    }));
+  }
+
+  /**
+   * Calculates what type of change, or none at all, the user made.
+   */
+  private _dataLimitChange(): Change {
+    if (this._showMenu) {
+      if (this._keyDataLimitBytes === undefined) {
+        // The user must have clicked the checkbox and input a limit.
+        return Change.SET;
+      }
+      const inputLimit = displayDataAmountToBytes(this.inputDataLimit());
+      if (inputLimit !== this._keyDataLimitBytes) {
+        return Change.SET;
+      }
+      return Change.UNCHANGED;
+    }
+    if (this._keyDataLimitBytes !== undefined) {
+      // The user must have unchecked the checkbox.
+      return Change.REMOVED;
+    }
+    return Change.UNCHANGED;
+  }
+
+  /**
+   * Returns true if the data limit was changed by the user.
+   */
+  public dataLimitChanged() {
+    return this._dataLimitChange() !== Change.UNCHANGED;
+  }
+
+  /**
+   * The current data limit as input by the user, but not necessarily as saved.
+   */
+  public inputDataLimit(): DisplayDataAmount {
```

Comment:
Get rid of this method and return the data limit value and unit in the `SavePerKeyDataLimitRequested` event instead.